Simplenote/SharedStorageMigrator.swift

```diff
@@ -0,0 +1,120 @@
+    /// Must run before Simperium is setup
+
+    func performMigrationIfNeeded() -> MigrationResult {
+        // Confirm if the app group DB exists
+        guard migrationNeeded else {
+            NSLog("Core Data Migration not required")
+            return .notNeeded
+        }
+
+        return migrateCoreDataToAppGroup()
+    }
+
+    private var migrationNeeded: Bool {
+        return legacyStorageExists && !sharedStorageExists
+    }
+
+    private func migrateCoreDataToAppGroup() -> MigrationResult {
+        NSLog("Database needs migration to app group")
+        NSLog("Beginning database migration from: \(storageSettings.legacyStorageURL.path) to: \(storageSettings.sharedStorageURL.path)")
+
+        do {
+            try disableJournaling()
+            try migrateCoreDataFiles()
+            try attemptCreationOfCoreDataStack()
+            NSLog("Database migration successful!!")
+            return .success
+        } catch {
+            NSLog("Could not migrate database to app group " + error.localizedDescription)
+            CrashLogging.logError(error)
+
+            removeFailedMigrationFilesIfNeeded()
+            return .failed
+        }
+    }
+
+    /// This method disables journaling on the core data database
+    /// Per this doc: https://developer.apple.com/library/archive/qa/qa1809/_index.html
+    /// Core data databases have journaling enabled by default, without disabling the journaling first it is possible some notes may get lost in migration
+    ///
+    private func disableJournaling() throws {
+        guard let mom = NSManagedObjectModel(contentsOf: storageSettings.modelURL) else {
+            throw NSError(domain: "SharedStorageMigrator", code: 100, userInfo: nil)
+        }
+        let psc = NSPersistentStoreCoordinator(managedObjectModel: mom)
+
+        let options = [NSSQLitePragmasOption: Constants.journalModeDisabled]
+
+        try psc.addPersistentStore(ofType: NSSQLiteStoreType,
+                                   configurationName: nil,
+                                   at: storageSettings.legacyStorageURL,
+                                   options: options)
+
+        // Remove the persistent store before exiting
+        // If removing fails, the migration can still continue so not throwing the errors
+        do {
+            for store in psc.persistentStores {
+                try psc.remove(store)
+            }
+        } catch {
+            NSLog("Could not remove temporary persistent Store " + error.localizedDescription)
+        }
+    }
+
+    private func migrateCoreDataFiles() throws {
+        try fileManager.copyItem(at: storageSettings.legacyStorageURL, to: storageSettings.sharedStorageURL)
+    }
+
+    private func attemptCreationOfCoreDataStack() throws {
+        NSLog("Confirming migrated database can be loaded at: \(storageSettings.sharedStorageURL)")
+        let _ = try CoreDataManager(storageSettings.sharedStorageURL)
```

Comment:
WDYT if weee

1. Rename `func disableJournaling` > `private func loadPersistentStorage(at storagePath: URL, journaling: Bool) throws`. And parametrize journaling + the Storage URL
2. Invoke `loadPersistentStorage(at: journaling)` from the two spots where `attemptCreationOfCoreDataStack`and `disableJournaling` are called now?

Only reason why doing this might be a good idea is that the current `disableJournaling` code **explicitly** removes the Persistent Store (and CoreDataManager does not!)

Reply:
Is it wasteful create `private func loadPersistentStorage(at storagePath: URL, journaling: Bool) throws `  as it's own method and then call it from  `func disableJournaling` and `attemptCreationOfCoreDataStack` (maybe change that to something like `verifyDataBaseLoadable`) and then in those methods print a few log messages

My thought here is that the migration becomes less readable to me like this:
    do {
            try loadPersistentStorage(at: URL, journaling: true)()
            try migrateCoreDataFiles()
            try loadPersistentStorage(at: URL, journaling: false)()
            NSLog("Database migration successful!!")
            return .success
}
compared to:
     do {
            try disableJournaling()
            try migrateCoreDataFiles()
            try attemptCreationOfCoreDataStack()
            NSLog("Database migration successful!!")
            return .success
}

Reply:
I made a quick update that shows what I was thinking with this.  WDYT?